<a href="https://colab.research.google.com/github/PolinaKokhaniukova/github-slideshow/blob/master/7_6_AB_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [34]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
#data = pd.read_csv('clickstrem.csv')
#data['timestamp'] = pd.to_datetime(data['InvoiceDate'])

In [73]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [39]:
import pandas as pd
df = pd.read_csv('drive/My Drive/clickstream.csv')

In [75]:
import csv
file_path = 'drive/My Drive/clickstream.csv'

In [76]:
with open(file_path, mode = 'r') as csv_file:
    csv_reader = csv.DictReader(csv_file, fieldnames = ['ID', 'page', 'date'])

In [77]:
data = pd.read_csv('drive/My Drive/clickstream.csv')

In [78]:
data.head(100)

client_id                      timestamp  ... card_type_used cart
0   796175212  2018-07-01 00:10:10.356057239  ...            NaN  NaN
1   796175212  2018-07-01 00:10:11.272860768  ...            NaN  NaN
2   796175212  2018-07-01 00:10:17.112819955  ...            NaN  NaN
3   796175212  2018-07-01 00:10:19.543343496  ...            NaN  NaN
4   419515020  2018-07-01 00:33:47.067460089  ...            NaN  NaN
..        ...                            ...  ...            ...  ...
95  304863183  2018-07-01 01:43:58.481779979  ...            NaN  NaN
96  304863183  2018-07-01 01:44:01.283568976  ...            NaN  NaN
97  304863183  2018-07-01 01:44:04.916105798  ...            NaN  NaN
98  304863183  2018-07-01 01:44:52.342786701  ...            NaN  NaN
99  680687203  2018-07-01 01:45:31.165645102  ...            NaN  NaN

[100 rows x 6 columns]

In [91]:
timestamp_ab = clickstream.query("timestamp >= '2019-09-02' & timestamp <= '2019-09-30'")
timestamp_ab.head()

client_id                      timestamp  ... card_type_used cart
2301173  294042545  2019-09-02 00:03:58.532762344  ...            NaN  NaN
2301174  294042545  2019-09-02 00:03:59.767865226  ...            NaN  NaN
2301175  294042545  2019-09-02 00:04:02.362878633  ...            NaN  NaN
2301176  134390951  2019-09-02 00:05:51.136832586  ...            NaN  NaN
2301177  134390951  2019-09-02 00:06:12.840263320  ...            NaN  NaN

[5 rows x 6 columns]

In [89]:
#data['client_id'].drop_duplicates()
data['client_id'].value_counts()

423140735    366
580223562    364
531284653    356
380932842    352
652981073    349
            ... 
742190523      2
679380470      2
651323498      2
349364347      2
750895160      2
Name: client_id, Length: 192182, dtype: int64

In [92]:
timestamp_ab['client_id'].value_counts()

205968525    137
189512331    137
228354857    116
612678224    103
730642869     98
            ... 
944729661      2
139831012      2
703969725      2
520964277      2
935942141      2
Name: client_id, Length: 16226, dtype: int64

In [42]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [93]:
import pandas as pd
df = pd.read_csv('drive/My Drive/ab_save_card_2019-09-02_2019-09-30.csv')

In [40]:
import csv
file_path = 'drive/My Drive/ab_save_card_2019-09-02_2019-09-30.csv'

In [41]:
data = pd.read_csv('drive/My Drive/ab_save_card_2019-09-02_2019-09-30.csv')

In [42]:
data.head()

client_id
0  294042545
1  134390951
2    9463642
3  818383160
4  433173698

In [43]:
data['client_id'].value_counts()

659473009    1
371181002    1
612464084    1
607927765    1
830137814    1
            ..
951157509    1
793496072    1
891466901    1
469650188    1
459032465    1
Name: client_id, Length: 8349, dtype: int64

Сравнила данные двух таблиц. Общее количество юзеров из первого списка 192182. При этом количество юзеров в период с 02.09.2019 по 30.09.2019 составило 16226. В группу для тестирования вошло 8349 участников, что составляет  51,4% от общего числа. Можно сказать, что группы для тестированя были достаточно одиноковыми по размеру.

In [44]:
ab_test = pd.read_csv('drive/My Drive/ab_save_card_2019-09-02_2019-09-30.csv')

In [45]:
clickstream = pd.read_csv('drive/My Drive/clickstream.csv')

Соединим две таблицы в одну, выберем все данные из общей талбицы по тем клиентам, которые были задействованы в AB тестировании.

In [46]:
pivot_ab_test_df = pd.merge(clickstream, ab_test, how='inner', on='client_id', right_index=True)


In [47]:
pivot_ab_test_df.columns
(['client_id', 'time_stamp', 'event', 'event_metadata',
       'card_type_used', 'cart'])

['client_id',
 'time_stamp',
 'event',
 'event_metadata',
 'card_type_used',
 'cart']

In [48]:
pivot_ab_test_df.head(100)

client_id  ...             cart
2770     406834357  ...              NaN
2771     406834357  ...              NaN
2772     406834357  ...              NaN
2773     406834357  ...              NaN
2774     406834357  ...              NaN
...            ...  ...              ...
1611513  406834357  ...  {'product5': 1}
1731056  406834357  ...              NaN
1731057  406834357  ...              NaN
1731058  406834357  ...              NaN
1731059  406834357  ...              NaN

[100 rows x 6 columns]

In [49]:
pivot_ab_test_df['client_id'].value_counts()

531284653    356
685699068    315
803952627    310
827949315    302
65927458     301
            ... 
362689474      2
947198433      2
259421444      2
998230516      2
208379680      2
Name: client_id, Length: 8349, dtype: int64

Попытаемся найти особенности группы, на основании которых она была выделена для тестирования.

In [50]:
pivot_ab_test_df['event'].value_counts()

products              88088
cart                  23662
lost                  21354
main                  19403
product5              19135
product3              15151
product2              10654
onboarding_main        8032
product1               7073
product4               5160
delivery_choice        3013
product3/cart/add      2901
payment_choice         1863
product2/cart/add      1501
payment_done           1371
delivery_pickup        1060
product1/cart/add      1038
product5/cart/add       982
delivery_courier        779
payment_card            769
payment_cash            705
input_card_success      677
product4/cart/add       276
payment_paypal          221
cart/delete             172
delivery_conv           105
Name: event, dtype: int64

In [1]:
!pip3 install 'retentioneering==2.0'

In [23]:
import retentioneering
import io
%matplotlib inline
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
import IPython
import pandas as pd
import numpy as np
import seaborn as sns


In [24]:
retentioneering.config.update({
    'event_col':'event',
    'event_time_col':'timestamp',
    'user_col': 'client_id'
})

In [25]:
data = 'pivot_ab_test_df'

In [13]:
IPython.display.HTML(data
 .retention
 .plot_graph(thresh=0.1, width=800, height=800))

AttributeError: ignored

In [26]:
IPython.display.HTML(data
 .rete
 .plot_graph(thresh=0.1, width=800, height=800))

AttributeError: ignored

In [27]:
IPython.display.HTML(
data.rete.plot_graph(thresh=0.01,
                     targets = {'payment_done':'green',
                                'lost':'red'}
                    )
)

AttributeError: ignored

In [16]:
data.rete.plot_graph(norm_type='full',
                     weight_col='client_id',
                     thresh=0.06,
                     targets = {'payment_done':'green',
                                'lost':'red'})


AttributeError: ignored

In [17]:
!pip show retentioneering

Name: retentioneering
Version: 2.0.0
Summary: Product analytics and marketing optimization framework based on deep user trajectories analysis
Home-page: https://github.com/retentioneering/retentioneering-tools
Author: Retentioneering User Trajectory Analysis Lab
Author-email: retentioneering@gmail.com
License: Retentioneering Software Non-Exclusive, Non-Commercial Use License (License)
Location: /usr/local/lib/python3.7/dist-packages
Requires: pymongo, umap-learn, vega, plotly, scipy, tqdm, matplotlib, altair, scikit-learn, networkx, seaborn, matplotlib, numpy, pandas
Required-by: 


In [19]:
df = data.rete.get_adjacency(weight_col='client_id', norm_type=None)

import seaborn as sns

#from IPython.html import widgets
from IPython.display import HTML
def rotateheader():
    return  [
      #  dict(selector="th.col_heading",
    #             props=[("writing-mode", "vertical-rl"), 
     #                   ('transform', 'rotateZ(-45deg)'),
     #                   ])
     dict(selector="th", props=[("font-size", "75%"),
                            ("text-align", "right")
                         
                          ]),
        dict(selector="th.col_heading", props=[("font-size", "75%"),
                            ("text-align", "center"),
                          ("transform", "translate(0%,-140%) rotate(-15deg)")
                          ]),
             dict(selector="caption", props=[("caption-side", "bottom")])
]
 

cm = sns.light_palette("green", as_cmap=True)
print("\n\n\n")
df.style.background_gradient(cmap=cm).set_table_styles(rotateheader()).set_caption(" .")

AttributeError: ignored

In [20]:
edgelist = data.rete.get_edgelist(weight_col='client_id', edge_attributes='users_nunique', norm_type=None)
edgelist.sort_values('users_nunique', ascending=False).head(5)

AttributeError: ignored

In [21]:
data.rete.get_clusters(plot_type='cluster_bar', n_clusters=10, targets = ['ayment_done'], ngram_range=(1, 1));

AttributeError: ignored